src/libraries/System.Private.CoreLib/src/System/IO/File.netcoreapp.cs

```diff
@@ -97,5 +103,108 @@ private static byte[] ReadAllBytesUnknownLength(FileStream fs)
                 }
             }
         }
+
+        private static void WriteToFile(string path, bool append, string? contents, Encoding encoding)
+        {
+            ReadOnlySpan<byte> preamble = encoding.GetPreamble();
+            int preambleSize = preamble.Length;
+            int maxFileSize = preambleSize + (string.IsNullOrEmpty(contents) ? 0 : encoding.GetMaxByteCount(contents.Length));
+            long preallocationSize = contents is null || contents.Length < ChunkSize ? 0 : maxFileSize; // for a single write operation setting preallocationSize has no perf benefit, as it requires additional sys-call(s)
+            FileMode mode = append ? FileMode.Append : FileMode.Create;
```

Comment:
Should the caller just pass in the FileMode rather than having the caller pass in a bool and branching on it here?